ludwig/models/abstractmodel.py

```diff
@@ -0,0 +1,242 @@
+import logging
+from abc import ABCMeta, abstractmethod
+from collections import OrderedDict
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+
+from ludwig.combiners.combiners import Combiner
+from ludwig.constants import COMBINED, LOSS, NAME, TIED, TYPE
+from ludwig.features.base_feature import InputFeature, OutputFeature
+from ludwig.features.feature_registries import input_type_registry, output_type_registry
+from ludwig.utils.algorithms_utils import topological_sort_feature_dependencies
+from ludwig.utils.metric_utils import get_scalar_from_ludwig_metric
+from ludwig.utils.misc_utils import get_from_registry
+from ludwig.utils.torch_utils import LudwigModule, reg_loss
+
+logger = logging.getLogger(__name__)
+
+
+class AbstractModel(LudwigModule, metaclass=ABCMeta):
```

Comment:
For an abstract class, there's a lot of default implementation.

Perhaps an overkill (and curious to hear others' thoughts), but a potentially more decoupled design would be to define a truly implementation-free abstract class `AbstractModel`, define a base class which defines several default implementations for several methods, `BaseModel`, and then have GBM and ECD based off of the `BaseModel`.